lib/api/core/vault.js

```diff
@@ -0,0 +1,218 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-console */
+
+'use strict';
+
+const
+  crypto = require('crypto'),
+  _ = require('lodash'),
+  fs = require('fs');
+
+const defaultSecretsFile = '../../../config/secrets.enc.json';
+
+class Vault {
+  static generateSeed () {
+    return crypto.randomBytes(8).toString('hex');
+  }
+
+  constructor (config) {
+    this.config = config;
+    this.secrets = {};
+  }
+
+  /**
+   * Prepare crypto primitives.
+   * Use the key passed in parameter or the KUZZLE_VAULT_KEY
+   * environment variable.
+   *
+   * @param {string|null} vaultKey - key used to decrypt the secrets
+   */
+  prepareCrypto (vaultKey = null) {
+    if (_.isString(vaultKey) && vaultKey.length > 0) {
+      this.vaultKey = vaultKey;
+    }
+    else if (_.isString(process.env.KUZZLE_VAULT_KEY)
+      && process.env.KUZZLE_VAULT_KEY.length > 0
+    ) {
+      this.vaultKey = process.env.KUZZLE_VAULT_KEY;
+    }
+    else {
+      return;
+    }
+
+    if (this.config.server.vaultSeed === '764c1afa2988981a') {
+      console.warn(
+        'You are using the default vault seed.\n',
+        'It\' STRONGLY advised to change this value.\n',
+        'Run the following commande to generate a new seed:\n',
+        '\tbin/kuzzle vaultSeed\n'
+      );
+    }
+
+    this.vaultKeyHash = crypto.createHash('md5')
```

Comment:
Never use md5, it's broken. If you need a cryptographic hash, use SHA512 (BLAKE2b is not supported in Node.js 6)

Why do we need to hash the provided key anyway? Is this to prevent storing the key in memory?

Reply:
In fact I does not need a cryptographic hash here, I need a key of at least 128 bits and must be a multiple of 32 bits. I only use md5 to ensure a correct size for the key but you're right, I can also use sha512

Reply:
ah ok, if that's the only purpose then we don't care